src/sentry/static/sentry/app/utils/bulkController.tsx

```diff
@@ -0,0 +1,134 @@
+import React from 'react';
+import xor from 'lodash/xor';
+import uniq from 'lodash/uniq';
+import intersection from 'lodash/intersection';
+
+import BulkNotice from 'app/components/bulkNotice';
+
+type RenderProps = {
+  /**
+   * Are all rows on current page selected?
+   */
+  isPageSelected: boolean;
+  /**
+   * Callback for toggling single row
+   */
+  onIdToggle: (id: string) => void;
+  /**
+   * Callback for toggling all rows across all pages
+   */
+  onAllIdsToggle: (select: boolean) => void;
+  /**
+   * Callback for toggling all rows on current page
+   */
+  onPageIdsToggle: (select: boolean) => void;
+  /**
+   * Ready to be rendered summary component showing how many items are selected,
+   * with buttons to select everything, cancel everything, etc...
+   */
+  bulkNotice: React.ReactNode;
+} & Pick<State, 'selectedIds' | 'isEverythingSelected'>;
+
+type State = {
+  /**
+   * Selected ids on the current page
+   */
+  selectedIds: string[];
+  /**
+   * Are all rows across all pages selected?
+   */
+  isEverythingSelected: boolean;
+};
+
+type Props = {
+  /**
+   * Array of ids on current page
+   */
+  pageIds: string[];
+  /**
+   * Number of all rows across all pages
+   */
+  allIdsCount: number;
+  /**
+   * Number of grid columns to stretch the selection summary (used in BulkNotice)
+   */
+  noticeColumns: number;
+  /**
+   * Children with render props
+   */
+  children: (props: RenderProps) => React.ReactNode;
+  /**
+   * Maximum number of rows that can be bulk manipulated at once (used in BulkNotice)
+   */
+  maxIdsCount?: number;
+};
+
+class BulkController extends React.Component<Props, State> {
```

Comment:
shouldn't this be in the components folder? Maybe we should discuss it in the next TSC meeting. I personally would expect to find more things like "lodash" functions in the utils and not components